posts/03-rust-packages-crates-modules.html.pm

```diff
@@ -0,0 +1,633 @@
+    Our code seemed correct, yet the metrics were not there.
+    It turned out that one of the packages used prometheus version ◊code{0.9} while all other packages used ◊code{0.10}.
+    According to ◊a[#:href "https://semver.org/"]{semver}, these versions are incompatible, so cargo linked both versions into the binary, introducing ◊em{two} implicit registries.
+    Only one of these implicit registries could be exposed.
+    The HTTP endpoint never pulled the metrics recorded to the other registry.
+  }
+  ◊p{
+    Passing loggers, metrics registries, async runtimes, etc. explicitly turns a runtime bug into a compile time error: the compiler will complain if you pass incompatible types around.
+    Switching to passing the metrics registry explicitly is what helped me to discover the issue with the metrics recording.
+  }
+  }
+}
+◊p{
+  The official documentation of the venerable ◊a[#:href "https://crates.io/crates/slog"]{slog} package also recommends passing loggers explicitly:
+}
+◊blockquote[#:cite "https://github.com/slog-rs/slog/wiki/FAQ#do-i-have-to-pass-logger-around"]{
+  ◊p{
+    The reason is: manually passing ◊code{Logger} gives maximum flexibility.
+    Using ◊code{slog_scope} ties the logging data structure to the stacktrace, which is not the same a logical structure of your software.
+    Especially libraries should expose full flexibility to their users, and not use implicit logging behaviour.
+  }
+  ◊p{
+    Usually ◊code{Logger} instances fit pretty neatly into data structures in your code representing resources, so it's not that hard to pass them in constructors, and use ◊code{info!(self.log, ...)} everywhere. 
+  }
+}
+◊p{
+ By passing state implicitly, you gain temporary convenience, but make your code less clear, less testable, and more error prone.
+ Every type of resource that we used to pass implicitly caused hard to diagnose issues (◊a[#:href "https://crates.io/crates/slog-scope"]{scoped} loggers, ◊a[#:href "https://crates.io/crates/prometheus"]{Prometheus} metrics registries, ◊a[#:href "https://crates.io/crates/rayon"]{Rayon} thread pools, ◊a[#:href "https://crates.io/crates/tokio"]{Tokio} runtimes, to name a few) and wasted a lot of engineering time.
+}
+◊p{
+  Some people in other programming communities also realized that global loggers are evil.
+  You might enjoy reading ◊a[#:href "https://www.yegor256.com/2019/03/19/logging-without-static-logger.html"]{Logging Without a Static Logger}, for example.
+}
+
+◊advice["dedup-dependencies"]{Deduplicate dependencies.}
+
+◊p{
+  Cargo makes it easy to add dependencies to your code, but it does not provide any tools to consolidate and maintain them in a large workspace.
+  At least until cargo developers implement ◊a[#:href "https://github.com/rust-lang/rfcs/pull/2906"]{RFC 2906}.
+  Until then, every time you bump a version of a dependency, try to do it consistently in all the packages in your workspace.
```

Comment:
Shouldn't we mention that one option is to use `cargo update` that will automatically try to update all dependencies in your `Cargo.lock` file?

Reply:
Great point, mentioned `cargo updated`.